beets/library.py

```diff
@@ -701,27 +701,32 @@ def move_file(self, dest, operation=MoveOperation.MOVE):
         """
         if not util.samefile(self.path, dest):
             dest = util.unique_path(dest)
-        if operation == MoveOperation.MOVE:
-            plugins.send("before_item_moved", item=self, source=self.path,
-                         destination=dest)
-            util.move(self.path, dest)
-            plugins.send("item_moved", item=self, source=self.path,
-                         destination=dest)
-        elif operation == MoveOperation.COPY:
-            util.copy(self.path, dest)
-            plugins.send("item_copied", item=self, source=self.path,
-                         destination=dest)
-        elif operation == MoveOperation.LINK:
-            util.link(self.path, dest)
-            plugins.send("item_linked", item=self, source=self.path,
-                         destination=dest)
-        elif operation == MoveOperation.HARDLINK:
-            util.hardlink(self.path, dest)
-            plugins.send("item_hardlinked", item=self, source=self.path,
-                         destination=dest)
-
-        # Either copying or moving succeeded, so update the stored path.
-        self.path = dest
+
+        try:
+            if operation == MoveOperation.MOVE:
+                plugins.send("before_item_moved", item=self, source=self.path,
+                             destination=dest)
+                util.move(self.path, dest)
+                plugins.send("item_moved", item=self, source=self.path,
+                             destination=dest)
+            elif operation == MoveOperation.COPY:
+                util.copy(self.path, dest)
+                plugins.send("item_copied", item=self, source=self.path,
+                             destination=dest)
+            elif operation == MoveOperation.LINK:
+                util.link(self.path, dest)
+                plugins.send("item_linked", item=self, source=self.path,
+                             destination=dest)
+            elif operation == MoveOperation.HARDLINK:
+                util.hardlink(self.path, dest)
+                plugins.send("item_hardlinked", item=self, source=self.path,
+                             destination=dest)
+        except FilesystemError as exc:
+            raise exc
+        else:
+            # Do not update library reference if filesystem operation failed
+            # Either copying or moving succeeded, so update the stored path.
+            self.path = dest
```

Comment:
Hmm, I think there may be something I'm not quite following here. Is there something that this new structure:

```
try:
  <the operation>
except:
  raise
else:
  <finalize>
```

accomplishes that the old `try`-less structure:

```
<the operation>
<finalize>
```

does not? It seems to me that, in both cases, the finalization step (self.path = dest) only gets executed when the move operation doesn't throw an exception. Is that right?